### PR TITLE
[space-picker] First iteration

### PR DESCRIPTION
**Pull request by DmitrySharabin**
Partially addresses #123.

### Usage examples

```html
<space-picker></space-picker>
```

```html
<space-picker value="oklab"></space-picker>
```

```html
<space-picker spaces="oklch, p3, srgb"></space-picker>
```

```html
<space-picker spaces="oklch, p3, srgb" value="p3"></space-picker>
```

```html
<space-picker onspacechange="this.nextElementSibling.textContent = this.value"></space-picker>
<output></output>
```

### CODE REVIEW

File: src/space-picker/space-picker.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+					}
+					catch (e) {
+						console.error(e);
+					}
+				},
+			},
+			default: () => Self.Color.spaces,
+			convert (value) {
+				// Drop non-existing spaces
+				let spaces = Object.fromEntries(Object.entries(value).filter(([id, space]) => space));
```
Comment (LeaVerou) [inline src/space-picker/space-picker.js:92]: I don't think we should drop them, I think we should just not list them with a name. Just the id.

Comment (LeaVerou) [reply src/space-picker/space-picker.js:92]: I.e. the name should default to the id.

Comment (DmitrySharabin) [reply src/space-picker/space-picker.js:92]: Makes sense. Will fix it. Thanks!

File: src/space-picker/space-picker.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+			},
+		},
+		spacechange: {
+			propchange: "value",
+		},
+	};
+
+	static formAssociated = {
+		like: el => el._el.picker,
+		role: "combobox",
```
Comment (LeaVerou) [inline src/space-picker/space-picker.js:119]: Doesn't this default to the element we specified in `like`? If not, it probably should.

File: src/space-picker/space-picker.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		valueProp: "value",
+		changeEvent: "spacechange",
```
Comment (LeaVerou) [inline src/space-picker/space-picker.js:121]: Shouldn't this use `change`?

File: src/space-picker/space-picker.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+				return spaces;
+			},
+			stringify (value) {
+				return Object.entries(value).map(([id, space]) => id).join(", ");
+			},
+		},
+	};
+
+	static events = {
+		change: {
```
Comment (LeaVerou) [inline src/space-picker/space-picker.js:107]: We should also retarget the `input` event.

File: src/space-picker/space-picker.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+				return Self.Color.Space.get(this._el.picker.value);
+			},
+			parse (value) {
+				if (value instanceof Self.Color.Space || value === null || value === undefined) {
+					return value;
+				}
+
+				value += "";
+
+				return Self.Color.Space.get(value);
```
Comment (LeaVerou) [inline src/space-picker/space-picker.js:67]: This is again the same issue with the string/object props. We need a prop that contains your actual input as a string (`value`) and one that contains the color space object (likely `space` or `selectedSpace`).

File: src/space-picker/space-picker.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		super.disconnectedCallback?.();
+		this._el.picker.removeEventListener("change", this);
+	}
+
+	handleEvent (event) {
+		if (event.type === "change" && event.target === this._el.picker) {
+			this.value = this._el.picker.value;
+		}
+
+		this.dispatchEvent(new event.constructor(event.type, {...event}));
```
Comment (LeaVerou) [inline src/space-picker/space-picker.js:31]: Doesn't defining the change event below take care of this?

**Comment by DmitrySharabin** [comment]
Thank you so much for your review! Every comment is on point, and I'll try to address it ASAP.

File: src/space-picker/space-picker.js (hunk 2 above)
Comment (DmitrySharabin) [reply src/space-picker/space-picker.js:119]: [In our code, we try to default to `computedRole`](https://github.com/nudeui/element/blob/3c3297eed2506200998b43e29eaa0512c1f0bdcc/src/formAssociated.js/defineFormAssociated.js#L35). Unfortunately, it still isn't supported. It's probably under the flag in some browsers (not sure).

image

**Comment by DmitrySharabin** [comment]
@LeaVerou, I addressed all your feedback (thank you for it). Here are the changes I made:

- Use `value` for actual input (as a string) and `selectedSpace` for the corresponding space object
- Support custom grouping of color spaces via the `groups()` prop
- Don't drop non-existent spaces; replace them with the `{ id, name: id }` objects
- Expose the `valuechange` and `spacechange` events
- Retarget the `input` event
- In `formAssociated`, use `change` as the `changeEvent`

Here is an example of using a custom grouping (from your color palettes research):

```html
<space-picker id="space_picker" spaces="oklch, p3, srgb" value = "p3"></space-picker>

<script type="module">
	space_picker.groups = (spaces) => {
		let ret = {polar: {}, rectangular: {}};
		for (let id in spaces) {
			let space = spaces[id];

			if (space.id !== id) {
				continue;
			}

			let isPolar = space.coords.h?.type === "angle";
			ret[isPolar ? "polar" : "rectangular"][id] = space;
		}
		return ret;
	};
</script>
```

image


Could you please take a look at this PR one more time?

File: src/space-picker/space-picker.js (hunk 2 above)
Comment (LeaVerou) [reply src/space-picker/space-picker.js:169]: We should just use `getComputedAccessibleNode()` instead. It's not like we don't have access to this information at all.

Comment (LeaVerou) [reply src/space-picker/space-picker.js:169]: (in Nude Element, in case that wasn't clear)

**Comment by DmitrySharabin** [comment]
Thank you so much for your feedback! I tried to address all the issues you pointed out. Namely:

* Add the `groupBy()` prop that expects a space and returns the name of a group the space belongs to

Providing grouping looks like so now:

```html
<space-picker id="space_picker" spaces="oklch, p3, srgb" value = "p3"></space-picker>

<script type="module">
	space_picker.groupBy = (space) => {
		let isPolar = space.coords.h?.type === "angle";
		return isPolar? "Polar" : "Rectangular";
	};
</script>
```

image


* Related to the previous one: `groups` is now a computed prop returning an object with all the groups
* Use different UI for cases with grouping and without

I hope I got everything right. If not, I'll be happy to fix it.

File: src/space-picker/space-picker.js (hunk 2 above)
Comment (DmitrySharabin) [reply src/space-picker/space-picker.js:181]: Got it. Thanks. I'll send a PR with the change.

**Review by LeaVerou — APPROVED** [review]
Looks good! Please add a README, and add it to the list of upcoming components.

**Comment by DmitrySharabin** [comment]
> Looks good! Please add a README, and add it to the list of upcoming components.

Thank you! Will do.

**Comment by DmitrySharabin** [comment]
Done! It's alive and can be played with: https://elements.colorjs.io/src/space-picker/ 🎉

**Pull request merged**